### PR TITLE
Garbage collection

### PR DESCRIPTION
**Pull request by awmatheson**

### CODE REVIEW

File: garbage-collection/gc_cleanup.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+def get_pods(workflow,namespace):
+    # check if pods exist in workflow and collect them
+    nodes = list(workflow['status']['nodes'].keys())
+    pods = []
+    for node in nodes:
+        try:
+            api_response = v1_api.read_namespaced_pod_status(node, namespace)
+            pods.append(node)
+        except ApiException as e:
+            # print("Exception when calling CoreV1Api->read_namespaced_pod: %s\n" % e)
```
Comment (inc0) [inline garbage-collection/gc_cleanup.py]: remove comments

File: garbage-collection/gc_cleanup.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+
+from kubernetes import client, config, watch
+from kubernetes.client.rest import ApiException
+from pprint import pprint
+
+
+# config.load_incluster_config()
+
+config.load_kube_config()
```
Comment (inc0) [inline garbage-collection/gc_cleanup.py]: switch to incluster

File: garbage-collection/gc_cleanup.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+            # print("Exception when calling CoreV1Api->read_namespaced_pod: %s\n" % e)
+            pass
+    return pods
+
+
+def delete_pods(pod,namespace,body):
+    try:
+        api_response = v1_api.delete_namespaced_pod(pod, namespace, body=body, propagation_policy='Background')
+    except ApiException as e:
+        print("Exception when calling CoreV1Api->delete_namespaced_pod: %s\n" % e)
```
Comment (inc0) [inline garbage-collection/gc_cleanup.py]: if exception is called, api response will not be initiated. Return will raise NameError. Also print that says workflow is deleted will be misleading

File: garbage-collection/gc_cleanup.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    workflows_not_expired = []
+    pods_deleted = []
+    for workflow in workflows['items']:
+        key = workflow['metadata']['name']
+
+        try:
+            finished_at = datetime.strptime(workflow['status']['finishedAt'], '%Y-%m-%dT%H:%M:%SZ')
+        except TypeError:
+            logging.info('could not read workflow {}'.format(key))
+        time_since_completion = (datetime.utcnow() - finished_at).total_seconds()/60/60
```
Comment (inc0) [inline garbage-collection/gc_cleanup.py:71]: again, if error is raised and caugth, `finished_at` will raise NameError

File: garbage-collection/gc_cleanup.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+                pods = get_pods(workflow,args.namespace)
+                for pod in pods:
+                    if not args.dry_run:
+                        delete_pods(pod,args.namespace,body)
+                    else:
+                        logging.info("dry_run flag set, would have deleted {}".format(pod))
+            else:
+                workflows_not_expired.append(key)
+        else:
+            exists = False
```
Comment (inc0) [inline garbage-collection/gc_cleanup.py]: this is repeated code with line 69, make func out of it?

File: garbage-collection/gc_cleanup.py [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+                # print("deleting workflow pod {} that completed {} hrs ago".format(key,time_since_completion))
+                pods = get_pods(workflow, args.namespace)
+                for pod in pods:
+                    if not args.dry_run:
+                        delete_pods(pod,args.namespace,body)
+                    else:
+                        logging.info("dry_run flag set, would have deleted {}".format(pod))
+            else:
+                workflows_not_expired.append(key)
+    logging.info("expired workflows: {}".format(workflows_expired))
```
Comment (inc0) [inline garbage-collection/gc_cleanup.py:97]: if we're using logging (we should!), let's replace all print statements with logging

File: garbage-collection/gc_cleanup.py [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+
+
+def main():
+  logging.basicConfig(level=logging.INFO)
```
Comment (inc0) [inline garbage-collection/gc_cleanup.py]: 4 space indetns plz

File: garbage-collection/gc_cleanup.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+# add in a group argument
+parser.add_argument("-grp", "--group", default="argoproj.io", type=str, help=("The custom resource's group name."))
+
+# add in a version argument
+parser.add_argument("-version", default="v1alpha1", type=str, help=("The custom resource's version."))
+
+# add in a plurals argument
+parser.add_argument("-p", "--plural", default="workflows", type=str, help=("The custom resource's plural name to filter by."))
+
+# add in a plurals argument
```
Comment (inc0) [inline garbage-collection/gc_cleanup.py]: this is an incorrect comment. Also, comments like this don't really add much information

File: garbage-collection/gc_cleanup.py [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+parser.add_argument("--max_age_hrs", default=168, type=int, help=("enter the maximum age to keep workflows for in hours"))
+
+parser.add_argument("--dry_run", action='store_true', help=("Triggers a dry run delete"))
+
+args = parser.parse_args()
+print(args)
+clean_up(args)
+
+if __name__ == "__main__":
+  main()
```
Comment (inc0) [inline garbage-collection/gc_cleanup.py]: 4 spaces plz

File: garbage-collection/garbage-collection-deployment.yaml [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+  name: hello
+spec:
+  schedule: "@weekly"
+  jobTemplate:
+    spec:
+      template:
+        spec:
+          containers:
+          - name: garbage-collection
+            image: argo-ml/latest
```
Comment (inc0) [inline garbage-collection/garbage-collection-deployment.yaml]: this image doesn't exist. We may want to publish images, but then it'd be something like github/argo-ml:latest.

**Review by inc0 — CHANGES_REQUESTED** [review]

File: garbage-collection/gc_cleanup.py (hunk 4 above)
Comment (awmatheson) [reply garbage-collection/gc_cleanup.py:71]: used continue to move to next workflow

**Pull request merged**